### DISCUSSION TITLE
Permissions in externals

### DISCUSSION THREAD

**Discussion by miquecg** [twpayne/chezmoi#2602] [Q&A] [answered]
Hi,

I've added font files to a `.chezmoiexternal` and now I'm getting this on the diff:

```
diff --git a/Library b/Library
old mode 40700
new mode 40755
diff --git a/Library/Fonts b/Library/Fonts
old mode 40700
new mode 40755
```

Fonts already exist on my system and folder permissions are OS defaults. Any idea or workaround to remove this discrepancy without changing folder permissions?

```
{{- $fontDir := "" }}
{{- if eq .chezmoi.os "darwin" }}
{{-   $fontDir = "Library/Fonts" }}
{{- else if eq .chezmoi.os "linux" }}
{{-   $fontDir = ".local/share/fonts" }}
{{- end }}
["{{ $fontDir }}/MesloLGS NF Regular.ttf"]
  type = "file"
  url = "https://github.com/romkatv/powerlevel10k-media/raw/master/MesloLGS NF Regular.ttf"
  refreshPeriod = "672h"
["{{ $fontDir }}/MesloLGS NF Bold.ttf"]
  type = "file"
  url = "https://github.com/romkatv/powerlevel10k-media/raw/master/MesloLGS NF Bold.ttf"
  refreshPeriod = "672h"
["{{ $fontDir }}/MesloLGS NF Italic.ttf"]
  type = "file"
  url = "https://github.com/romkatv/powerlevel10k-media/raw/master/MesloLGS NF Italic.ttf"
  refreshPeriod = "672h"
["{{ $fontDir }}/MesloLGS NF Bold Italic.ttf"]
  type = "file"
  url = "https://github.com/romkatv/powerlevel10k-media/raw/master/MesloLGS NF Bold Italic.ttf"
  refreshPeriod = "672h"
```

**Comment by twpayne** [comment, marked as answer]
This is occurring because the directories `~/Library` and `~/Library/Fonts` are implicitly created by `.chezmoiexternal` and implicitly created directories have default (i.e. non-private permissions).

The fix is to tell chezmoi to manage the directories explicitly. This can be done with:

```console
$ chezmoi add --recursive=false ~/Library ~/Library/Fonts
```

You'll likely need to add `.keep` file to the directory in the source state, otherwise git will ignore it:

```console
$ touch $(chezmoi source-path ~/Library/Fonts)/.keep
```

**Reply by miquecg** [reply]
Well, had to finally add those two folder manually.

```
❯ chezmoi add --recursive=false ~/Library ~/Library/Fonts
chezmoi: rename /Users/me/.local/share/chezmoi/Library/Fonts /Users/me/.local/share/chezmoi/Library/private_Fonts: no such file or directory
```

It also happened to me when trying to add a file in `.oh-my-zsh/custom` because it also is an implicitly managed folder by chezmoi. I found this was an [issue](https://github.com/twpayne/chezmoi/issues/1614) for other people too so I followed advice there and commented out `exact = true` in `.chezmoiexternal` for OhMyZsh. But now looking at your own personal dotfiles I tried putting it back and adding `exact_` in front of both folder names: `exact_dot_oh-my-zsh/exact_custom`, which works.

By the way, do I have to conditionally add `Library/Fonts` and `.local/share/fonts` to `.chezmoiignore` now that they need to be explicitly added?

**Reply by miquecg** [reply]
Thanks for this great tool 🙂

**Reply by twpayne** [reply]
> By the way, do I have to conditionally add `Library/Fonts` and `.local/share/fonts` to `.chezmoiignore` now that they need to be explicitly added?

Yes. If you don't then they'll be automatically created on every machine.

**Reply by twpayne** [reply]
> By the way, do I have to conditionally add `Library/Fonts` and `.local/share/fonts` to `.chezmoiignore` now that they need to be explicitly added?

This is only needed if you have other files in `Library/Fonts`/`.local/share/fonts` managed by chezmoi outside your `.chezmoiexternal` file. If they're only mentioned in your `.chezmoiexternal` then you don't need to add them to `.chezmoiignore`.
